micropython/upip/upip.py

```diff
@@ -0,0 +1,173 @@
+
+
+def _install_json(package_json_url, index, target, version, mpy):
+    response = requests.get(_rewrite_url(package_json_url, version))
+    try:
+        if response.status_code != 200:
+            print("Package not found:", package_json_url)
+            return False
+
+        package_json = response.json()
+    finally:
+        response.close()
+    for target_path, short_hash in package_json.get("hashes", ()):
+        fs_target_path = target + "/" + target_path
+        if _check_exists(fs_target_path, short_hash):
+            print("Exists:", fs_target_path)
+        else:
+            file_url = "{}/file/{}/{}".format(index, short_hash[:2], short_hash[2:])
+            if not _download_file(file_url, fs_target_path):
+                print("File not found: {} {}".format(target_path, short_hash))
+                return False
+    for target_path, url in package_json.get("urls", ()):
+        fs_target_path = target + "/" + target_path
+        if not _download_file(_rewrite_url(url, version), fs_target_path):
+            print("File not found: {} {}".format(target_path, url))
+            return False
+    for dep, dep_version in package_json.get("deps", ()):
+        if not _install_package(dep, index, target, dep_version, mpy):
+            return False
+    return True
+
+
+def _install_lib(package, index, target, version, mpy):
+    if not version:
+        version = "latest"
+    print("Installing {}@{} from {} to {}".format(package, version, index, target))
+
+    mpy_version = (
+        sys.implementation._mpy & 0xFF if mpy and hasattr(sys.implementation, "_mpy") else "py"
+    )
+
+    package_json_url = "{}/package/{}/{}/{}.json".format(index, mpy_version, package, version)
+    return _install_json(package_json_url, index, target, version, mpy)
+
+
+def _install_package(package, index, target, version, mpy):
+    if (
+        package.startswith("http://")
+        or package.startswith("https://")
+        or package.startswith("github:")
+    ):
+        if package.endswith(".py") or package.endswith(".mpy"):
+            print("Downloading {} to {}".format(package, target))
+            return _download_file(
+                _rewrite_url(package, version), target + "/" + package.rsplit("/")[-1]
+            )
+        else:
+            if not package.endswith(".json"):
+                if not package.endswith("/"):
+                    package += "/"
+                package += "package.json"
+            print("Installing {} to {}".format(package, target))
+            return _install_json(package, index, target, version, mpy)
+    else:
+        return _install_lib(package, index, target, version, mpy)
+
+
+def install(package, index=_PACKAGE_INDEX, target=None, version=None, mpy=True):
+    if not target:
+        for p in sys.path:
+            if p == "lib" or p.endswith("/lib"):
+                target = p
+                break
+        else:
+            print("Unable to find lib dir in sys.path.")
```

Comment:
no other message ends in a full-stop in this file, so i don't think this one should (and saves a byte!)